api-examples/transformer_utils.py

```diff
@@ -0,0 +1,422 @@
+from baseline.pytorch.torchy import vec_log_sum_exp
+from baseline.pytorch.seq2seq import Seq2SeqModel
+from eight_mile.utils import str2bool, write_yaml, read_yaml, Offsets
+from eight_mile.pytorch.layers import *
+import baseline.pytorch.embeddings
+import baseline.embeddings
+from baseline.progress import create_progress_bar
+from torch.utils.data.dataset import IterableDataset, TensorDataset
+from baseline.vectorizers import Token1DVectorizer, BPEVectorizer1D
+import codecs
+from collections import Counter
+import glob
+
+
+class TripletLoss(nn.Module):
+    """Provide a Triplet Loss using the reversed batch for negatives"""
+    def __init__(self, model):
+        super().__init__()
+        self.score = nn.CosineSimilarity(dim=1)
+        self.model = model
+
+    def forward(self, inputs, targets):
+        # reverse the batch and use as a negative example
+        neg = targets.flip(0)
+        query = self.model.encode_query(inputs)
+        response = self.model.encode_response(targets)
+        neg_response = self.model.encode_response(neg)
+        pos_score = self.score(query, response)
+        neg_score = self.score(query, neg_response)
+        score = neg_score - pos_score
+        score = score.masked_fill(score < 0.0, 0.0).sum(0)
+        return score
+
+
+class AllLoss(nn.Module):
+    def __init__(self, model):
+        r"""Loss from here https://arxiv.org/pdf/1705.00652.pdf see section 4
+
+        We want to minimize the negative log prob of y given x
+
+        -log P(y|x)
+
+        P(y|x) P(x) = P(x, y)                             Chain Rule of Probability
+        P(y|x) = P(x, y) / P(x)                           Algebra
+        P(y|x) = P(x, y) / \sum_\hat(y) P(x, y = \hat(y)) Marginalize over all possible ys to get the probability of x
+        P_approx(y|x) = P(x, y) / \sum_i^k P(x, y_k)      Approximate the Marginalization by just using the ys in the batch
+
+        S(x, y) is the score (cosine similarity between x and y in this case) from our neural network
+        P(x, y) = e^S(x, y)
+
+        P(y|x) = e^S(x, y) / \sum_i^k e^S(x, y_k)
+        log P(y|x) = log( e^S(x, y) / \sum_i^k e^S(x, y_k))
+        log P(y|x) = S(x, y) - log \sum_i^k e^S(x, y_k)
+        -log P(y|x) = -(S(x, y) - log \sum_i^k e^S(x, y_k))
+        """
+        super().__init__()
+        self.c = 1.0
+        self.score = nn.CosineSimilarity(dim=-1)
+        self.model = model
+        self.max_scale = math.sqrt(self.model.embedding_layers.get_dsz())
```

Comment:
From the paper it says `computed as cosine similarity with annealing between the encodings hx and hy. It starts at 1 and ends at √d` so it seems like this should actually be the sqrt of the size of the inputs rather than the size of the embeddings. I know that our transformers don't really change sizes as data flows trough but I think this seems unsafe

Reply:
the input size in this code is guaranteed to be the same between embeddings and hidden units.